cpp/include/utilities/cython.hpp

```diff
@@ -109,6 +110,60 @@ struct graph_container_t {
   experimental::graph_properties_t graph_props;
 };
 
+// replacement for std::tuple<,,>, since std::tuple is not
+// supported in cython
+//
+template <typename vertex_t, typename weight_t>
+struct major_minor_weights_t {
+  explicit major_minor_weights_t(raft::handle_t const& handle)
+    : shuffled_major_vertices_(0, handle.get_stream()),
+      shuffled_minor_vertices_(0, handle.get_stream()),
+      shuffled_weights_(0, handle.get_stream())
+  {
+  }
+  rmm::device_uvector<vertex_t>& get_major(void) { return shuffled_major_vertices_; }
```

Comment:
No, because they're used by `std::tie()` to assign the fields directly.

Reply:
That feels awkward and non-intuitive.  That might be worth either reconsidering (although I'm not entirely sure there's a more natural idiom for what is required) or documenting explicitly so that it doesn't surprise someone maintaining the code later.

Reply:
I'm not sure what exactly "feels awkward" for you. We have a non-const getter that allows direct assignment of a field, thereby avoiding copies. If that struct was an aggregate the fields would have been assigned directly. 

I don't know what "feels intuitive" either. These qualifications seem mighty subjective to me. Can you recommend a concrete course of action, here?

Reply:
FYI, see https://github.com/aschaffer/cugraph/blob/47f7aff07aca087e6f6f6fedc7b4bd78c177aedb/cpp/src/experimental/coarsen_graph.cu#L337 .

It's the same thing here, except the fields are private, hence they need accessors by reference (the reasons that they're private: (1) better encapsulation; (2) this struct is exposed to cython, which theoretically sees the entire public interface, but these fields cannot be exposed to cython; hence they're private)

Reply:
Clearly subjective...

I guess I'm accustomed to seeing a struct aggregate and directly assigning the fields when you're going to be assigning the fields in this way.  When I see accessors returning the fields then I expecting an RAII style object or something that more resembles an RAII style object.

Perhaps this is a common idiom that I just haven't seen much of.  I just worry if it's not a common idiom that it will be confusing to people just glancing at it.

Reply:
std::tie(a, b) = some_func();

a & b needs to be lvalue references, but some_func() does not need to return a reference type. sort_and_shuffle_values in the above link return rvalue (lvalue does not work as copy constructor is deleted in rmm::device_uvector).

Reply:
@seunghwak , if you have a sfunction, similar to `sort_and_shuffle()`:
```
template<typename T>
auto fv1(void)
{
  std::vector<T> v{6, 7, 8, 9};
  return std::move(v);
}
```
(or, returning `decltype(auto)`, which in general is more recommended), then the return type is `vector<T>`:
```
//... stuff...
  auto ret_f1 = fv1<double>();

  static_assert(std::is_same<decltype(ret_f1), std::vector<double> >::value,"Error: type mismatch...");
```
The above static assertion passes. 

And in cases like that `std::tie()` works as expected ("ties" the lvalue reference to the return). Example:
```
template <typename T, typename U>
std::tuple<std::vector<T>, U, T> foo(void)
{
  std::vector<T> v{1, 2, 3, 4, 5};
  U u{13};
  T t{17};
  return std::make_tuple(v, u, t);
}

template <typename T, typename U>
struct X {
  X(void) = default;

  std::vector<T>& get_v(void) { return v_; }
  U& get_u(void) { return u_; }
  T& get_t(void) { return t_; }

 private:
  std::vector<T> v_;
  U u_;
  T t_;
};

int main(void)
{
  X<int, double> x{};
  std::tie(x.get_v(), x.get_u(), x.get_t()) = foo<int, double>(); // assigns return of foo() to underlying X fields
}
```

So, I'm not sure what is the issue here? And what is the desired behavior?

Reply:
If I change `foo()`'s signature to look more like `sort_and_shuffle()`, the outcome is the same:
```
template <typename T, typename U>
auto foo(void)
{
  std::vector<T> v{1, 2, 3, 4, 5};
  U u{13};
  T t{17};
  return std::make_tuple(std::move(v), u, t);
}
```

Reply:
So, you're using .get_major() in the left side of the statement, something like

major_minor_weights_t obj{};
std::tie(obj.get_major(), obj.get_minor(), obj.get_weights()) = a_func_that_returns_a_tuple();

I thought you're trying to use this in the right side. I don't have a big complaint about this, but related to what @ChuckHastings said, we may consider few other options.

1. If we're using struct to just tie some variables, should the struct member variables be private? Directly exposing a non-const reference to a private member variable sounds non-idiomatic to me.

2. Another option is to add a setter function taking a tuple of majors, minors, weights as an input parameter if you really want to keep the struct member variables as private.

Reply:
And do we have any guidelines on when to use struct vs class? I use struct to just simply tie variables (variables are public in this case) and class if I need to limit direct access to the member variables and truly need some class member functions.

Reply:
`get_major()` etc. are returning refences. For all intent and purpose...they can be used as either getters or setters.

Reply:
BTW, @seunghwak, what you're doing in coarsen_graph.cu:L337 ( https://github.com/aschaffer/cugraph/blob/47f7aff07aca087e6f6f6fedc7b4bd78c177aedb/cpp/src/experimental/coarsen_graph.cu#L337) is
```
    std::forward_as_tuple(
      std::tie(rx_edgelist_major_vertices, rx_edgelist_minor_vertices, rx_edgelist_weights),
      std::ignore) =
      sort_and_shuffle_values(...);
```
Which seems much less "intuitive" than what I do. I'm not sure why that is okay and my code above isn't.

In the end I'd like to resolve this conversation, which with all due respect doesn't seem to go anywhere. To that end, what is the desired change here? If none, can we resolve this?

Reply:
So, the code above, it should ideally be std::tie(std::tie(), ...), but std::tie does not take rvalue reference, so std::forward_as_tuple replaces the outer std::tie (std::forward_as_tuple is nearly identical to std::tie in this context, but it takes a rvalue reference). People who are not familiar with std::forward_as_tuple may find the code above confusing, but it is just using std::tie in a nested way.

My biggest concern about the code above is, allowing users to directly modify the private member variable (by returning a non-const reference). Then, why define those variables as private from the first place?

If we are using "struct major_minor_weights_t" to just tie variables, then we may just do

```
struct major_minor_weights_t {
  rmm::device_uvector<vertex_t> shuffled_major_vertices;  <= public
  rmm::device_uvector<vertex_t> shuffled_minor_vertices;
  rmm::device_uvector<weight_t> shuffled_weights;
};

major_minor_weights_t ob{}j;
std::tie(obj.shuffled_major_vertices, obj.shuffled_minor_vertices, shuffled_weights) = func();
```

instead of defining the member variables as private but also provide getter functions for direct modification.

But this is just my suggestion, and I don't think this is a deal breaker, so I am not against marking this conversation as resolved.

Reply:
Because, as already explained: 
(1) This gets exposed to Cython; Cython "sees" public interfaces; if you make those public we might have (if not right away, possibly in the future) issues with exposing a type (`device_uvector<>`) that Cython does _not_ understand (yet);
(2) data fields in classes should _always_ be private (it's in Scott Meyers idioms, it's in the code review guidelines, it's...good practice);

Also, as already noted you also use a getter to set something (a tuple element); again: 
```
tests/experimental/coarsen_graph_test.cpp:196:            std::get<0>(cur) = cugraph::experimental::invalid_vertex_id<vertex_t>::value;
```
If semantically this practice seems wrong to you...why did you use it yourself?

Resolving this.